inspirehep/modules/disambiguation/core/ml/sampling.py

```diff
@@ -0,0 +1,86 @@
+# along with INSPIRE. If not, see <http://www.gnu.org/licenses/>.
+#
+# In applying this license, CERN does not waive the privileges and immunities
+# granted to it by virtue of its status as an Intergovernmental Organization
+# or submit itself to any jurisdiction.
+
+"""Disambiguation core ML sampling."""
+
+from __future__ import absolute_import, division, print_function
+
+import itertools
+import json
+import random
+from collections import defaultdict
+
+import six
+
+
+def sample_signature_pairs(signatures_path, clusters_path):
+    blocks = defaultdict(list)
+    signatures_reversed = {}
+    with open(signatures_path, 'r') as fd:
+        for line in fd:
+            signature = json.loads(line)
+            blocks[signature['signature_block']].append(signature['signature_uuid'])
+            signatures_reversed[signature['signature_uuid']] = signature
+
+    clusters_reversed = {}
+    with open(clusters_path, 'r') as fd:
+        for line in fd:
+            cluster = json.loads(line)
+            for signature_uuid in cluster['signature_uuids']:
+                clusters_reversed[signature_uuid] = cluster
+
+    same_cluster_same_name = []
+    same_cluster_different_name = []
+    different_cluster_same_name = []
+    different_cluster_different_name = []
+    for _, block in six.iteritems(blocks):
+        for s1, s2 in itertools.combinations(block, 2):
+            s1_cluster = clusters_reversed[s1]['cluster_id']
+            s2_cluster = clusters_reversed[s2]['cluster_id']
+            s1_name = signatures_reversed[s1]['author_name']
+            s2_name = signatures_reversed[s2]['author_name']
+            if s1_cluster == s2_cluster and s1_name == s2_name:
+                same_cluster_same_name.append({'same_cluster': True, 'signature_uuids': [s1, s2]})
+            elif s1_cluster == s2_cluster and s1_name != s2_name:
+                same_cluster_different_name.append({'same_cluster': True, 'signature_uuids': [s1, s2]})
+            elif s1_cluster != s2_cluster and s1_name == s2_name:
+                different_cluster_same_name.append({'same_cluster': False, 'signature_uuids': [s1, s2]})
+            elif s1_cluster != s2_cluster and s1_name != s2_name:
+                different_cluster_different_name.append({'same_cluster': False, 'signature_uuids': [s1, s2]})
+
+    random.shuffle(same_cluster_same_name)
+    random.shuffle(same_cluster_different_name)
+    random.shuffle(different_cluster_same_name)
+    random.shuffle(different_cluster_different_name)
+    for el in roundrobin(
+        itertools.cycle(same_cluster_same_name),
+        itertools.cycle(same_cluster_different_name),
+        itertools.cycle(different_cluster_same_name),
+        itertools.cycle(different_cluster_different_name),
+    ):
+        yield el
+
+
+def roundrobin(*iterables):
+    nexts = itertools.cycle(iter(it).next for it in iterables)
+    for next in nexts:
+        yield next()
```

Comment:
This comes from the [itertools recipes](https://docs.python.org/2/library/itertools.html#recipes).